forwardServer.go

```diff
@@ -0,0 +1,269 @@
+		os.Exit(1)
+	}
+
+	go func() {
+		for {
+			conn, ok := <-connChan
+			if !ok {
+				sock.Close()
+				break
+			}
+			conn.Reject()
+		}
+	}()
+
+	for {
+		conn, err := sock.Accept()
+		if err != nil {
+			log.Warning("Failed to accept connection from os", err)
+			break
+		}
+
+		var addr string
+		var port uint32
+		var agentCon io.ReadWriteCloser
+		switch setup.Protocol {
+		case "tcp":
+			addr = conn.RemoteAddr().(*net.TCPAddr).IP.String()
+			port = uint32(conn.RemoteAddr().(*net.TCPAddr).Port)
+			agentCon, err = forwardCtx.NewConnectionTCP(
+				setup.BindHost,
+				setup.BindPort,
+				addr,
+				uint32(port),
+				func() error {
+					log.Warning("Close callback, closing conn")
+					return conn.Close()
+				},
+			)
+		case "unix":
+			agentCon, err = forwardCtx.NewConnectionUnix(
+				setup.BindHost,
+				func() error {
+					log.Warning("Close callback, closing conn")
+					return conn.Close()
+				},
+			)
+		default:
+			panic(fmt.Errorf("Unknown protocol %s", setup.Protocol))
+		}
+		if err != nil {
+			log.Warning("Failed to create new connection with backend", err)
+		}
+
+		go serveConnection(log, conn, agentCon)
+		go serveConnection(log, agentCon, conn)
+
+		if setup.SingleConnection {
+			break
+		}
+	}
+
+	log.Debug("Waiting for connections to finish")
+
+	forwardCtx.WaitFinish()
+}
+
+func externalDial(log log.Logger, forwardCtx *proto.ForwardCtx, connChan chan *proto.Connection, setup proto.SetupPacket) {
+    defer func() {
+        if x := recover(); x != nil {
+            log.Error(fmt.Sprintf("Panicked with %+v", x))
+        }
+    }()
+	for {
+		agentCon, ok := <-connChan
+		if !ok {
+			break
+		}
+		details := agentCon.Details()
+		protocol := "tcp"
+		if details.Protocol == proto.PROTOCOL_UNIX {
+			protocol = "unix"
+		}
+		log.Warning(fmt.Sprintf("Dialing %s %s:%d", setup.Protocol, details.ConnectedAddress, details.ConnectedPort))
+
+		dialAddr := parsePort(protocol, details.ConnectedAddress, details.ConnectedPort)
+
+		conn, err := net.Dial(protocol, dialAddr)
+		if err != nil {
+			log.Warning("Failed to dial %s", dialAddr, err)
+			agentCon.Reject()
+			continue
+		}
+		agentCon.Accept()
+		go serveConnection(log, conn, agentCon)
+		go serveConnection(log, agentCon, conn)
+	}
+
+	forwardCtx.WaitFinish()
+}
+
+func forwardServer(stdin io.Reader, stdout io.Writer, stderr io.Writer, exit exitFunc) {
+	logConfig := config.LogConfig{
+		Level:       config.LogLevelDebug,
+		Destination: config.LogDestinationFile,
+		File:        "/tmp/agent.log",
+		Format:      "ljson",
+	}
+	log, err := log.NewLogger(logConfig)
+	if err != nil {
+		panic(err)
+	}
+
+	sigc := make(chan os.Signal, 1)
+	sig2.Notify(sigc,
+		syscall.SIGINT,
+		syscall.SIGTERM,
+		syscall.SIGQUIT)
+	go func() {
+		for {
+			_ = <-sigc
+			// ... do something ...
+			buf := make([]byte, 1<<20)
+			stacklen := runtime.Stack(buf, true)
+			str := fmt.Sprintf("=== received SIGQUIT ===\n*** goroutine dump...\n%s\n*** end\n", buf[:stacklen])
```

Comment:
We should think about a good way to serialize these in the log library. Logging this kind of stuff may prove tricky with syslog if we ever decide to add it.